### ISSUE TITLE
FR: opt-out from openvpn-CCD generation

### ISSUE THREAD

**Issue by EugenMayer** [opnsense/core#2319]
I would provide a patch for this, just wanted to be sure there is interest.

I would love to opt-out from openvpn-based CCD generation ( legacy CCD ) which is also based on the old xml structure.

Since i have writte https://github.com/EugenMayer/opnsense-openvpn-plugin and use the state-of-the-art model based serialisation, i get into troubles when e.g. a openvpn server configuration is saved - since this regenerates the CCDs using the old method ( basically in my case, removing all CCDs ).

Can i offer a checkbox, in the advanced section of an openvpn server, to opt-out from automatic CCD handling? The point is, if someone decides to generated the CCD using a different logic, the legacy openvpn plugin will always removed the CCDs, and then replace it with "nothing" (since there will be no legacy ones ).

So the point of that feature is: opt-out from openvpn CCD generation, in the that case "automatic removal". Would be just a checkbox.

Of course, default is opt-in, so no current setups will changed or anything. 

Any interest? ( i`ll do the work)

**Comment by AdSchellevis** [comment]
is this still for parsing the client config via radius? using the` Framed-* ` attributes are likely quite easy to fix and prevents adding different hooks for standard use-cases (https://github.com/opnsense/plugins/pull/485#issuecomment-357460986)

**Comment by EugenMayer** [comment]
No actually this has nothing to do with radius at all - its basically a rewrite of the legacy CCD support with the new API + adding REST to it

**Comment by mimugmail** [comment]
Would it be possible to set a different CCD directory via Advanced and put your plugin output files there?

**Comment by AdSchellevis** [comment]
The best solution probably is to extend (refactor) https://github.com/opnsense/core/blob/master/src/etc/inc/plugins.inc.d/openvpn/auth-user.php at some point to be able to write out the client configuration after successful authentication and offer an option to plug in custom hooks there, that way it's more aligned to how others use this  too (e.g. https://github.com/grayscott/radiusplugin/blob/453df3493c7aa11b132bd9b6104e7291ecdd2d7d/AuthenticationProcess.cpp#L95)

**Comment by EugenMayer** [comment]
@AdSchellevis well rewriting the openvpn plugin to be non legacy is for sure the best option, bet that is a very long road, with a a huge codebase change affecting a lot of current installation - this is not going to happen any soon, nor getting merged any sooner nor even getting released prio 19.x i would suppose.

So this is an intermediate solution.

The radius plugin is already non-legacy, it does not really help comparing it to the current state of the legacy openvpn plugin.

So i guess there is a little chance this effort would get merged, i am i right?

**Comment by AdSchellevis** [comment]
I'm not saying we should replace the whole openvpn component, just improving the authentication sequence, which isn't an awful lot of work by the way.
Adding new hooks, which only serve code hosted somewhere else are usually not accepted, because it increases maintenance and user errors which are hard to track.

Maybe we can leave this issue open, if I can find some time somewhere (no promises), I might try to fix this, since the radius use-case would benefit as well from it.

**Comment by EugenMayer** [comment]
@AdSchellevis i was not asking for a hook here, i understand, that this ends up in spaghetti code. 

Let me rephrase, i want to offer a checkbox, which is then used in the opevpn legacy CCD code generation, if checked, its not generated. And thats it. So its a 2 line if statement in the CCD code.

I would put it here https://github.com/opnsense/core/blob/master/src/etc/inc/plugins.inc.d/openvpn.inc#L1076 and thats about it.

Adding a checkbox to the GUI + storage if that is the rest of the PR

**Comment by AdSchellevis** [comment]
I understand, just don't think that's the correct approach for the issue. The issue seems broader then disabling the csc generation, automatically leading to new hooks and exceptions at a later point in time.

This probably needs some work to fix it properly, which will return at some point again (most likely radius auth, but more use-cases probably exist).

**Comment by EugenMayer** [comment]
@AdSchellevis ok, i cannot see how this raises any expectation at all - this is the non integrated "i deal with it myself" kind of way, with a flag to disable the core handling. The opposite of any hooks, expectations or anything.

I guess i have to hot-patch my system for now then. are patches by opnsense-patch actually preserved over updates?

**Comment by fabianfrz** [comment]
@EugenMayer no, they will be reverted by the next update. opnsense-patch is to add changes, before the next release is out. The main purpose is for testing new features and use it for hotfixes.

**Comment by AdSchellevis** [comment]
Don't get me wrong, we deal with a huge code base, we always have to ask ourselves how changes pin out for others, what the expected maintenance and support fallback might be in combination with planned or unplanned future changes.
In the end, we have to deal (and pay) for the maintenance of most of them in the core system.

I don't mind leaving the issue open, as stated earlier, since it would be a good idea to fix the generation of client configs at some point. But that's up to you.

**Comment by EugenMayer** [comment]
@fabianfrz thanks, makes sense to any extend. So i have to care about that myself, including upstream merges so the patch anchors stay put. Thanks for the info.

Closing this issue since it has nothing to do with "fixing CCD / CSC generation", it is for disabling this. @AdSchellevis closing it on my behalf - if you feel like it makes sense to switch the focus of this issue and reopen it, feel free to go.
